services/secure-gateway/security/matchResource.go

```diff
@@ -0,0 +1,102 @@
+package security
+
+import (
+	"fmt"
+	"strings"
+)
+
+// MatchResources is used as a wrapper for matchResourcesFunc for casbin registration
+func MatchResources(args ...interface{}) (interface{}, error) {
+	requestResource := args[0]
+	policyResource := args[1]
+	rr, ok := requestResource.(map[string]string)
+	if !ok {
+		return nil, fmt.Errorf("Expected map[string]string, but got %T (%v)", requestResource, requestResource)
+	}
+
+	pr, ok := policyResource.(string)
+	if !ok {
+		return nil, fmt.Errorf("Expected string, but got %v (%T)", requestResource, requestResource)
+	}
+
+	return matchResourcesFunc(rr, pr)
+}
+
+// matchResourcesFunc parses the policy resource (pr) and verifies that it matches what was given in the request (rr)
+func matchResourcesFunc(rr map[string]string, pr string) (bool, error) {
+	parsedPolicyResource, err := parseResource(pr)
+	if err != nil {
+		return false, err
+	}
+
+	if len(parsedPolicyResource) != len(rr) {
+		return false, nil
+	}
+
+	for key, value := range parsedPolicyResource {
+		resourcePart, ok := rr[key]
+		if !ok { // some keys are missing from the request - no match
+			return false, nil
+		}
+		if !matchWithWildcards(value, resourcePart) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func matchWithWildcards(pattern, input string) bool {
+	// any
+	if pattern == "*" {
+		return true
+	}
+
+	// prefix
+	if strings.HasSuffix(pattern, "*") {
+		return strings.HasPrefix(input, pattern[0:len(pattern)-2])
+	}
+
+	// no wildcards
+	return pattern == input
+}
+
+func parseResource(resource string) (result map[string]string, err error) {
+	// split by `:`
+	parts := strings.Split(resource, ":")
+	result = make(map[string]string)
+
+	// empty string
+	if len(resource) == 0 {
+		return result, nil
+	}
+
+	ctxs := []string{}
```

Comment:
add const for key, not very readable in current form

Reply:
consider making "parseResource" return (ctxs, key, error).
and reading key/path from matchResourcesFunc at the start.
It might make the algorithm more easy to read and it'll separate "prefix" logic with "any" logic